### PR TITLE
Add a new feature flag enable_virtual_delete_vms, when it was enabled…

### PR DESCRIPTION
**Pull request by jianqiu**
When instance upgrade, bosh director will first delete vm from infrastructure then recreate the vm. Under some special conditions, for example, in SoftLayer, we don`t want to real delete vm from Softlayer, bosh softlayer cpi would leverage the os_reload provided by Softlayer to upgrade the stemcell/instance vms.
So, I just submit this PR which including a new feature flag to enable virtual delete vm when instance upgrade.

### CODE REVIEW

**Comment by jianqiu** [comment]
@cppforlife @maximiliem @mattcui
Hi Dmitriy, 
When I submitted the PR, the travis job failed due to bundler version, could you please help me look into it? Thanks!

travis_time:end:02463ab0:start=1462334179980041597,finish=1462334183268746566,duration=3288704969
[0Ktravis_fold:end:install.1
[0Ktravis_fold:start:install.2
[0Ktravis_time:start:247772c4
[0K$ bundle install --local --without development
Resolving dependencies...
[33mSome gems seem to be missing from your vendor/cache directory.[0m
[31mBundler could not find compatible versions for gem "bundler":
  In Gemfile:
    bundler (~> 1.11.0)

  Current Bundler version:
    bundler (1.12.1)
This Gemfile requires a different version of Bundler.
Perhaps you need to update Bundler by running `gem install bundler`?

Could not find gem 'bundler (~> 1.11.0)' in any of the sources[0m

**Comment by jianqiu** [comment]
@cppforlife @maximiliem @mattcui
Hi Dimitriy, could you please help me to look at the travis ci failure? Thanks!

**Comment by dpb587-pivotal** [comment]
@jianqiu, sorry for the confusion - the Travis build failures are unrelated to your changes.

**Comment by MichaelTrestman** [comment]
Hi! @jianqiu, thank you for the pr.

The unit tests fail when we run them. We think that [this line](https://github.com/jianqiu/bosh/blob/af46cb76319d4298c9cc22c1818b7d2def1609b7/bosh-director/lib/bosh/director/vm_deleter.rb#L31) could be the problem. Please confirm that all unit tests pass and then push again, and we'll take another look.

Thanks for contributing,
Michael and @Kiemes

**Comment by jianqiu** [comment]
@MichaelTrestman, @cppforlife, @maximiliem, @mattcui

Hi, Michael, thanks for your suggestion. After modified the codes, I still got the following error, in fact, I am not sure whether my changes impacted these cases. could you please help me look into the travis failure? Thanks for your kind help!


rspec ./spec/unit/db/migrations/director/20160427164345_add_teams_spec.rb:10 # add_teams migrates deployment teams over to Teams table
rspec ./spec/unit/db/migrations/director/20160427164345_add_teams_spec.rb:23 # add_teams preserves the teams associated with a deployment using a many-to-many table
rspec ./spec/unit/db/migrations/director/20160427164345_add_teams_spec.rb:37 # add_teams should check that deployments_teams has unique deployment_id and team_id pairs
rspec ./spec/unit/db/migrations/director/20160427164345_add_teams_spec.rb:47 # add_teams removes the teams column from deployment
/home/travis/build/cloudfoundry/bosh/bosh-dev/lib/bosh/dev/tasks/spec.rake:100:in `unit_exec'
/home/travis/build/cloudfoundry/bosh/bosh-dev/lib/bosh/dev/tasks/spec.rake:138:in `block (6 levels) in <top (required)>'
/home/travis/build/cloudfoundry/bosh/bosh_common/lib/common/thread_pool.rb:77:in `block (2 levels) in create_thread'
/home/travis/build/cloudfoundry/bosh/bosh_common/lib/common/thread_pool.rb:63:in `loop'
/home/travis/build/cloudfoundry/bosh/bosh_common/lib/common/thread_pool.rb:63:in `block in create_thread'
/home/travis/.rvm/gems/ruby-2.3.1/gems/logging-1.8.2/lib/logging/diagnostic_context.rb:323:in `block in create_with_logging_context'
Tasks: TOP => spec:unit:ruby

**Comment by jianqiu** [comment]
@cppforlife @MichaelTrestman @maximiliem @mattcui
Is there any update to this PR? Thanks!

**Comment by jianqiu** [comment]
@cppforlife @maximilien @mattcui
Could you please look into it? Thanks!

**Comment by allomov** [comment]
@jianqiu I afraid, that the team can be aware of the some technical debt for this work around. This can be my personal feeling, but maybe such kind of feature should be re-thinked in some way.

**Comment by xoebus** [comment]
We (@aminjam and myself) just ran the tests again for this PR and everything is looking green. This is getting near being merged. We have a few questions about some of the changes that we will comment inline below.

Thanks!

File: bosh-director/spec/unit/instance_deleter_spec.rb [hunk 1]
```diff
@@ -273,6 +273,25 @@ module Bosh::Director
             end
           end
         end
+
+        context 'when virtual_delete_vm option is passed in' do
+          let(:options) { {virtual_delete_vm: true} }
+
+          it 'deletes snapshots, persistent disk, releases old reservations, vm should not delete from cloud' do
```
Comment (xoebus) [inline bosh-director/spec/unit/instance_deleter_spec.rb]: I'm able to completely remove the implementation that this test corresponds with and this test still passes. Are you sure it's testing what you want it to test?

File: bosh-director/lib/bosh/director/vm_deleter.rb [hunk 2]
```diff
@@ -26,7 +28,9 @@ def delete_for_instance(instance)
 
     def delete_vm(vm_cid)
       @logger.info('Deleting VM')
-      @error_ignorer.with_force_check { @cloud.delete_vm(vm_cid) }
+      @error_ignorer.with_force_check do
+        @cloud.delete_vm(vm_cid) unless @enable_virtual_delete_vm
```
Comment (xoebus) [inline bosh-director/lib/bosh/director/vm_deleter.rb]: *

File: bosh-director/lib/bosh/director/vm_creator.rb [hunk 3]
```diff
@@ -156,7 +156,7 @@ def create(instance_model, stemcell, cloud_properties, network_settings, disks,
       @logger.error("error creating vm: #{e.message}")
       if vm_cid
         parent_id = add_event(instance_model.deployment.name, instance_model.name, 'delete', vm_cid)
-        @vm_deleter.delete_vm(vm_cid)
+        @vm_deleter.delete_vm(vm_cid) unless Config.enable_virtual_delete_vms
```
Comment (xoebus) [inline bosh-director/lib/bosh/director/vm_creator.rb]: I think there are tests missing for this conditional behavior.

**Comment by xoebus** [comment]
After these things are addressed it would also be nice if the commits here could be [squashed into a single commit][squash].

[squash]: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

**Comment by cppforlife** [comment]
@jianqiu we are going to take a stab at fixing tests.

**Comment by jianqiu** [comment]
@cppforlife  Thanks for your kind help!

**Comment by jianqiu** [comment]
@cppforlife @mattcui @maximilien

Any update?  Thanks!

**Comment by jianqiu** [comment]
I saw the contents of this PR had been merged into develop branch, your help is greatly appreciated, and thanks for helping me make some corrections to my codes. So I am closing this PR.

**Pull request closed without merging**